lib/email/receiver.rb

```diff
@@ -668,11 +669,10 @@ def subject
         end
     end
 
-    def find_or_create_user(email, display_name, raise_on_failed_create: false)
-      user = nil
-
+    def find_or_create_user(email, display_name, raise_on_failed_create: false, user: nil)
       User.transaction do
-        user = User.find_by_email(email)
+        user ||= User.find_by_email(email)
+        @staged_users << user if user&.staged?
```

Comment:
How about moving this at the end of the transaction so we don't have to add users twice?

```ruby
User.transaction do
  user ||= User.find_by_email(email)
   
  if user.nil? && SiteSetting.enable_staged_users
    user = create_staged_users
    @created_staged_users << user
  end
   
  @staged_users << user if user&.staged?
end
```